core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -629,6 +692,23 @@ private void sendUpdate()
                 dynamicFilterDomains.getDynamicFilterDomains(),
                 session.getExchangeEncryptionKey());
         byte[] taskUpdateRequestJson = taskUpdateRequestCodec.toJsonBytes(updateRequest);
+
+        if (enableAdaptiveTaskRequestSize && taskUpdateRequestJson.length > maxRequestSize) {
```

Comment:
Testing here `taskUpdateRequestJson.length > maxRequestSize` disallow to increase `currentSplitBatchSize` if request size is smaller than `expectedSize` and we didn't exceed  `maxUnacknowledgedSplits`. Cannot we have `taskUpdateRequestJson.length > (maxRequestSize - headroom)` in if at 704 line?

Also, it would be nice to test increment behavior.